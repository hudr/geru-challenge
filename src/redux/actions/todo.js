import API from '../../service'

import {
  GET_TODOS,
  ADD_TODO,
  UPDATE_TODO,
  REMOVE_TODO,
  CLOSE_MODAL,
  OPEN_MODAL
} from './types'

export function getTodos() {
  return dispatch => {
    API.get(`gerutodo`)
      .then(res => {
        dispatch({
          type: GET_TODOS,
          payload: res.data
        })
      })
      .catch(err => {
        console.log(err)
      })
  }
}

export function addTodo(data) {
  return dispatch => {
    API.post(`gerutodo`, data)
      .then(res => {
        dispatch({
          type: ADD_TODO,
          payload: res.data
        })
      })
      .catch(err => {
        console.log(err)
      })
  }
}

export function updateTodo(data) {
  return dispatch => {
    API.put(`gerutodo/${data.id}`, data)
      .then(res => {
        dispatch({
          type: UPDATE_TODO,
          payload: res.data
        })
      })
      .catch(err => {
        console.log(err)
      })
  }
}

export function removeTodo(todo) {
  return dispatch => {
    API.delete(`gerutodo/${todo}`)
      .then(() => {
        dispatch({
          type: REMOVE_TODO,
          payload: todo
        })
      })
      .catch(err => {
        console.log(err)
      })
  }
}

export function openModal(type, currentTodo = {}) {
  return dispatch => {
    dispatch({
      type: OPEN_MODAL,
      payload: { opened: true, type, currentTodo }
    })
  }
}

export function closeModal() {
  return dispatch => {
    dispatch({
      type: CLOSE_MODAL,
      payload: false
    })
  }
}
