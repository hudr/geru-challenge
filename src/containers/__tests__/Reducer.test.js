import reducer from '../../redux/reducers/todo'
import * as types from '../../redux/actions/types'

describe('todos reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual({
      todos: [],
      isModalOpen: false,
      modalType: '',
      currentTodo: {}
    })
  })
})

describe('Covering Actions', () => {
  const expectedState = {
    todos: [],
    modalType: '',
    currentTodo: {},
    isModalOpen: false
  }

  test('ADD_TODO', () => {
    const action = {
      type: types.ADD_TODO,
      payload: {
        id: '2',
        createDate: '2019-09-14T03:45:16.709Z',
        title: 'Badwork',
        description: 'Description',
        dueDate: '2019-09-14T03:45:16.709Z',
        estTime: '2',
        tags: ['grow', 'up'],
        done: 'false'
      }
    }
    const changedState = {
      ...expectedState,
      todos: [...expectedState.todos, action.payload],
      isModalOpen: false
    }

    expect(reducer(undefined, action)).toEqual(changedState)
    expect(reducer(undefined, action)).toMatchSnapshot(changedState)
  })

  test('UPDATE_TODO', () => {
    const action = {
      type: types.UPDATE_TODO,
      payload: {
        id: '1',
        createDate: '2019-09-14T03:45:16.709Z',
        title: 'Badwork',
        description: 'Description',
        dueDate: '2019-09-14T03:45:16.709Z',
        estTime: '2',
        tags: ['grow', 'up'],
        done: 'false'
      }
    }
    const changedState = {
      ...expectedState,
      todos: [...expectedState.todos, action.payload],
      isModalOpen: false
    }

    expect(reducer(undefined, action)).toMatchSnapshot(changedState)
  })

  test('GET_TODOS', () => {
    const action = {
      type: types.GET_TODOS,
      payload: {
        id: '1',
        createDate: '2019-09-14T03:45:16.709Z',
        title: 'Homework',
        description: 'Description',
        dueDate: '2019-09-14T03:45:16.709Z',
        estTime: '2',
        tags: ['grow', 'up'],
        done: 'false'
      }
    }
    const changedState = { ...expectedState, todos: action.payload }
    expect(reducer(undefined, action)).toEqual(changedState)
    expect(reducer(undefined, action)).toMatchSnapshot(changedState)
  })

  test('REMOVE_TODOS', () => {
    const action = {
      type: types.REMOVE_TODO,
      payload: 1
    }
    const changedState = {
      ...expectedState,
      todos: expectedState.todos.filter(todo => {
        return todo.id !== action.payload
      })
    }

    expect(reducer(undefined, action)).toEqual(changedState)
    expect(reducer(undefined, action)).toMatchSnapshot(changedState)
  })

  test('CLOSE_MODAL', () => {
    const action = { type: types.CLOSE_MODAL, payload: false }
    const changedState = { ...expectedState, isModalOpen: action.payload }
    expect(reducer(undefined, action)).toEqual(changedState)
    expect(reducer(undefined, action)).toMatchSnapshot(changedState)
  })

  test('OPEN_MODAL', () => {
    const action = {
      type: types.OPEN_MODAL,
      payload: { opened: true, type: '', currentTodo: {} }
    }
    const changedState = {
      ...expectedState,
      isModalOpen: action.payload.opened,
      modalType: action.payload.type,
      currentTodo: action.payload.currentTodo
    }
    expect(reducer(undefined, action)).toEqual(changedState)
    expect(reducer(undefined, action)).toMatchSnapshot(changedState)
  })
})
