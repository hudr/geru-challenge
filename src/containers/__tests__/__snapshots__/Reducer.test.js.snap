// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Covering Actions ADD_TODO 1`] = `
Object {
  "currentTodo": Object {},
  "isModalOpen": false,
  "modalType": "",
  "todos": Array [
    Object {
      "createDate": "2019-09-14T03:45:16.709Z",
      "description": "Description",
      "done": "false",
      "dueDate": "2019-09-14T03:45:16.709Z",
      "estTime": "2",
      "id": "2",
      "tags": Array [
        "grow",
        "up",
      ],
      "title": "Badwork",
    },
  ],
}
`;

exports[`Covering Actions CLOSE_MODAL 1`] = `
Object {
  "currentTodo": Object {},
  "isModalOpen": false,
  "modalType": "",
  "todos": Array [],
}
`;

exports[`Covering Actions GET_TODOS 1`] = `
Object {
  "currentTodo": Object {},
  "isModalOpen": false,
  "modalType": "",
  "todos": Object {
    "createDate": "2019-09-14T03:45:16.709Z",
    "description": "Description",
    "done": "false",
    "dueDate": "2019-09-14T03:45:16.709Z",
    "estTime": "2",
    "id": "1",
    "tags": Array [
      "grow",
      "up",
    ],
    "title": "Homework",
  },
}
`;

exports[`Covering Actions OPEN_MODAL 1`] = `
Object {
  "currentTodo": Object {},
  "isModalOpen": true,
  "modalType": "",
  "todos": Array [],
}
`;

exports[`Covering Actions REMOVE_TODOS 1`] = `
Object {
  "currentTodo": Object {},
  "isModalOpen": false,
  "modalType": "",
  "todos": Array [],
}
`;

exports[`Covering Actions UPDATE_TODO 1`] = `
Object {
  "currentTodo": Object {},
  "isModalOpen": false,
  "modalType": "",
  "todos": Array [
    Object {
      "createDate": "2019-09-14T03:45:16.709Z",
      "description": "Description",
      "done": "false",
      "dueDate": "2019-09-14T03:45:16.709Z",
      "estTime": "2",
      "id": "1",
      "tags": Array [
        "grow",
        "up",
      ],
      "title": "Badwork",
    },
  ],
}
`;
